### ISSUE TITLE
Poor performance with multi-GPU

### ISSUE THREAD

**Issue by hfutxrg** [tensorflow/tensorflow#4744]
I tried to train the example inception-v3 network on multiple GPUs. The following commands are used:
/* use 4 GPUs */
bazel-bin/inception/imagenet_train --num_gpus=4 --batch_size=64 --max_steps=2000 --train_dir=... --data_dir=...

/*use 2 GPUs */
export CUDA_VISIBLE_DEVICES=0,1
bazel-bin/inception/imagenet_train --num_gpus=2 --batch_size=32 --max_steps=2000 --train_dir=... --data_dir=...

/*use 1 GPU */
export CUDA_VISIBLE_DEVICES=0
bazel-bin/inception/imagenet_train --num_gpus=1 --batch_size=16 --max_steps=2000 --train_dir=... --data_dir=...

The performance results are as follows:
GPUs   Training time(s)       samples/sec
1           657                          52.7
2           844                          97.3
4           1104                        150

The training time was got from the log file by subtracting the first time from the last time and therefore excluded the start-up time. For example
2016-10-03 15:28:23.239148: step 0, loss = 13.08 (4.6 examples/sec; 13.916 sec/batch)
....
2016-10-03 15:46:47.830959: step 1990, loss = 12.21 (174.6 examples/sec; 0.367 sec/batch)

Now the part that I do not understand is that the samples/sec scales normally with increasing number of GPUs, but the training time does not reduce. Instead the training also increases with more GPUs. 

Has anyone observed the similar case? 

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
I tried to set --input_queue_memory_factor=0 as in the post https://github.com/tensorflow/models/issues/47, but it does not help.

Also the post https://github.com/tensorflow/tensorflow/issues/4272 has similar performance issue, but it was not solved.

### Environment info
Operating System: Redhat 7.2

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
-rw-r--r-- 1 root root   560184 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudadevrt.a
lrwxrwxrwx 1 root root       16 Sep 12 15:13 /usr/local/cuda-8.0/lib64/libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root root       19 Sep 12 15:13 /usr/local/cuda-8.0/lib64/libcudart.so.8.0 -> libcudart.so.8.0.27
-rwxr-xr-x 1 root root   394472 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudart.so.8.0.27
-rw-r--r-- 1 root root   737516 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudart_static.a
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so.5
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so.5.1.3
-rw-r--r-- 1 root root 59715990 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn_static.a

If installed from source, provide 
1. The commit hash (`git rev-parse HEAD`): a63b0cbcabc79531e155a0663a08656debf2fe07
2. The output of `bazel version`: 

.
Build label: 0.3.1- (@non-git)
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Sep 29 22:19:27 2016 (1475187567)
Build timestamp: 1475187567
Build timestamp as int: 1475187567

### Logs or other output that would be helpful
By the way, before the actual computation, the overhead of the initialization took >7minutes, is that normal? And I also got many warnings:
WARNING:tensorflow:tf.op_scope(values, name, default_name) is deprecated, use tf.name_scope(name, default_name, values)
WARNING:tensorflow:tf.variable_op_scope(values, name, default_name) is deprecated, use tf.variable_scope(name, default_name, values)
WARNING:tensorflow:Passing a `GraphDef` to the SummaryWriter is deprecated. Pass a `Graph` object instead, such as `sess.graph`.

Will all these warnings impact the performance?

**Comment by tatatodd** [comment]
@hfutxrg have you seen the following documentation:
https://github.com/tensorflow/models/tree/master/inception
https://github.com/tensorflow/models/tree/master/inception#how-to-train-from-scratch

In particular the following section may be useful:

---

A crucial aspect of training a network of this size is training speed in terms of wall-clock time. The training speed is dictated by many factors -- most importantly the batch size and the learning rate schedule. Both of these parameters are heavily coupled to the hardware set up.

Generally speaking, a batch size is a difficult parameter to tune as it requires balancing memory demands of the model, memory available on the GPU and speed of computation. Generally speaking, employing larger batch sizes leads to more efficient computation and potentially more efficient training steps.

We have tested several hardware setups for training this model from scratch but we emphasize that depending your hardware set up, you may need to adapt the batch size and learning rate schedule.

Please see the comments in inception_train.py for a few selected learning rate plans based on some selected hardware setups.

**Comment by hfutxrg** [comment]
@tatatodd I did read those pages. But based on my understanding, the training time (wall-clock time) is negatively correlated with samples/sec. If samples/sec increases, then the training time should decrease. 

Did you mean there is no direct relationship between training time and samples/sec?

**Comment by tatatodd** [comment]
@hfutxrg I see your point, and don't have an answer.  Perhaps @poxvoculi can shed some light on this?

**Comment by poxvoculi** [comment]
The phenomenon reported is interesting.  It's possible there's a bug in the form of a computational inaccuracy somewhere, but it also seems possible (and without further evidence more likely) that this is a natural consequence of SGD learning and the hyper-parameters you've chosen.  In particular, as you've increased the number of GPUs, you've also increased the batch size linearly.  

There is not a simple, direct relationship between samples/sec and training time.

It's not the case that 1 batch of 64 is likely to provide equivalent progress to the learning goal as 4 batches of 16.  The learning path from initialized parameters to fully-trained parameters is through a high-dimensional space, and likely very twisty.  Computing the gradient from a larger mini-batch gives a higher expectation of a good approximation of the true gradient at a given point, but doesn't generally justify a much larger learning step, i.e. a larger learning rate, because moving just a little bit can change the true gradient, regardless of whether one's approximation from the last point was accurate.

Picking a minibatch size is a pragmatic compromise, best done empirically.  Suppose for the sake of argument that computing the gradient for a minibatch of size k takes k times longer than for a single sample, and that the expected goodness of the gradient improves by 1/(k+1) whenever you increase the minibatch size by 1.  In this case it seems pretty clear that the best SGD minibatch size is 1, because the expected advantage in gradient accuracy is worth less than being able to take steps faster.  In practice, especially with SIMD devices, the time to compute a minibatch is much less than k times larger than a simple sample, so the best minibatch size will usually be larger than 1.

Consider your edge cases of batch sizes 64 and 16.  At 150 samples per sec, the large batch was 0.43 sec/step, and at 52 samples per sec the small batch was 0.31 secs/step, correct?   Using the small batch, your reported training time was 657 secs or 2120 steps.  Using the large batch the training time was 1104 sec or 2700 steps.   So, the larger batch in this case seems worse, both because the wall clock time is longer, and because the number of steps to completion is larger.

By the preceding argument, I'm less concerned with the inverse relation between samples/sec and training time than with steps/sec and training time.   What determines when training terminates?  Are you using the same learning rate in both cases?   It strikes me as rather suspicious that you're able to train a model in < 3000 steps, is this just a retraining?

You might try varying only the number of GPUs, without varying the batch size, to see what effect it has on steps to completion.  Make sure you process the same inputs, in the same order.

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity. Please reopen when additional information becomes available. Thanks!

**Comment by shashank879** [comment]
I am also experiencing a similar issue. I get almost same total training time while training my network on a single GPU (TitanX 12GB) and 4 GPUs(1080 8GB each), and selecting a proper batch_size via trial and error, single GPU outperforms the 4 GPUs. I anyone is willing to look into it I can provide the necessary details.
